### ISSUE TITLE
debugger not working properly with the latest vscode 1.60.2 and the "julia 1.4.3" extension. 

### ISSUE THREAD

**Issue by MariusDrulea** [julia-vscode/julia-vscode#2465]
When I debug, the "All" package list in always empty. I have vscode 1.60.2 and the "julia 1.4.3" extension.  Two weeks ago everything was running fine but something got updated in the background.

![image](https://user-images.githubusercontent.com/27725572/134508630-aca9349b-6c24-4f9f-9627-f2b7947f5662.png)

**Comment by pfitzseb** [comment]
That list is only populated when a Julia REPL is running.

**Comment by MariusDrulea** [comment]
which is therefore the recommended way to debug? 
1. Run with debugging (F5) - it starts a Julia Debugger console
2. start a REPL and start the debugging session here

**Comment by pfitzseb** [comment]
Either way is fine, though I'd usually recommend not using a separate process. That's also not really relevant to the symbol list you were talking about originally -- nothing's stopping you from just having a REPL session running alongside the separate debugger process or whatever. And you also don't really need this "All" list since you can add/remove functions or modules by clicking the `+` button.

**Comment by MariusDrulea** [comment]
I am currently playing with option1. I have manually set all the packages to be "compiled" except for "Main". "Main.Base" and "Main.Core" are set to "compile" as well. _Strictly my scripts are set to be interpreted_. 

The effect is that the debugger only stops at the run_normals() function, see above picture. When I "Step in" all the breakpoints are skipped, the debugger no longer stops. Am I missing something?
[settings.json.txt](https://github.com/julia-vscode/julia-vscode/files/7218208/settings.json.txt)

**Comment by pfitzseb** [comment]
Well, breakpoints don't work in compiled code.
Can't say more without knowing your code though.

**Comment by MariusDrulea** [comment]
I have create a very simple code for this. I want to debug strictly my_fun and nothing else.  Except for Main.my_fun which is "interpreted", eveything else is set to "compile", including Main.Base and Main.Core. The debugger is able to stop at line 7. However it does not step inside my_fun.
![image](https://user-images.githubusercontent.com/27725572/134532620-52a6da11-b077-4f28-a248-1039a27bd1f8.png)

**Comment by MariusDrulea** [comment]
@pfitzseb : thanks for this quick fix.
I would like to contribute with testing, as debugging is a really important feature to have for julia.
Once the fix is merged into master, can I test it with julia-insider + regular vscode (not vscode-insider)?

**Comment by pfitzseb** [comment]
There's no fix, I tagged this issue by accident.
Generally I wouldn't recommend setting Main as compiled, and definitely not all of Base.

**Comment by MariusDrulea** [comment]
I noticed if I set All->Main.Base as compile and All->Main.Core as interpreted, everything works as expected. However, when I run my larger code, debugging is slow in the following code, second line.

```
wide_arr = Matrix{UInt16}(undef, stride, height)
img_data[:, :] = Matrix{UInt8}(wide_arr[:, 1:width] .>> 4)   # debugger is very slow in this line
```

There was this broadcasting operation which was set to "interpreted" in Compile->Base. Set it to compile and the debugger is able to quickly execute that line.

**Comment by pfitzseb** [comment]
Yeah, that's a matter of configuration, basically. Broadcasting is interpreted by default because it's often used with user-defined functions (much like `map` and similar higher order functions).

**Comment by MariusDrulea** [comment]
We have Base and Core in the following locations:
1. Compiled->Base, Compiled->Core
2. All->Base, All->Core
3. All->Main.Base, All->Main.Core

Q1: Is there a priority list for 1, 2 and 3? E.g. a setting in 3 overrides the same setting in 1. 

Q2: I am able to edit the "Compiled" list via the text file in "C:\Users\ddr5clj\AppData\Roaming\Code\User\settings.json".  Is there a similar file for the "All" list? it is a lot easier to edit this list via txt file.

The reason behind Q2 is that I have some for-loops in my larger code which are very slow. I need to set _lots_ of All->Main.Core to "interpreted" (e.g. Int16, toInt, Vector etc)

**Comment by MariusDrulea** [comment]
For Q2, it's possible to set the desired "Main.Core.Int64" and many others manually in the IDE and then save the current configuration as default. The _setting.json_ file will contain these items. 

I am still trying to figure out which of the Main.Core items have to stay "interpreted" such that we can step inside functions.

**Comment by pfitzseb** [comment]
The default is to compile everything in Base except for the functions and modules explicitly listed as interpreted, so you should never need to set a bunch of functions as compiled yourself.

Q1: No, no special order. One of these definitions will win out, but it's basically undefined which one.
Q2: No, that's runtime generated. And you don't need to use it at all anyways.

**Comment by MariusDrulea** [comment]
Bug inside debugger? I have noticed when I set All->Main.Core.Vararg to "compile", the debugger crashes with the following error:
```
 Error: Some Julia code in the VS Code extension crashed
└ @ VSCodeDebugger c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\error_handler.jl:15
ERROR: TypeError: in Type, in parameter, expected Type, got Vararg{T, N}
Stacktrace:
  [1] signature_type(f::Any, args::Any)
    @ Base .\reflection.jl:793
  [2] _methods
    @ .\reflection.jl:856 [inlined]
  [3] methods(f::Any, t::Any, mod::Nothing)
    @ Base .\reflection.jl:912
  [4] methods (repeats 2 times)
    @ .\reflection.jl:935 [inlined]
  [5] set_compiled_functions_modules!(items::Vector{String})
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:157
  [6] set_compiled_functions_modules!
    @ c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:173 [inlined]
  [7] set_compiled_items_request(conn::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, state::VSCodeDebugger.DebugAdapter.DebuggerState, params::NamedTuple{(:compiledModulesOrFunctions,), Tuple{Vector{String}}})
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:99
  [8] debug_notification(conn::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, state::VSCodeDebugger.DebugAdapter.DebuggerState, params::VSCodeDebugger.DebugAdapter.DebugArguments)
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:41
  [9] (::VSCodeDebugger.DebugAdapter.var"#120#146"{VSCodeDebugger.DebugAdapter.DebuggerState})(conn::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, params::VSCodeDebugger.DebugAdapter.DebugArguments)
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\packagedef.jl:49
 [10] dispatch_msg(x::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, dispatcher::VSCodeDebugger.JSONRPC.MsgDispatcher, msg::Dict{String, Any})
    @ VSCodeDebugger.JSONRPC c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\JSONRPC\src\typed.jl:67    
 [11] macro expansion
    @ c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\packagedef.jl:76 [inlined]       
 [12] (::VSCodeDebugger.DebugAdapter.var"#143#169"{VSCodeDebugger.var"#3#4"{Tuple{String, String}}, VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, VSCodeDebugger.JSONRPC.MsgDispatcher})()
    @ VSCodeDebugger.DebugAdapter .\task.jl:411

Julia debuggee finished. Press ENTER to close this terminal.
```

**Comment by MariusDrulea** [comment]
Weird thing:
1. If I individually set _all the items in All->Main.Core.... to "compile"_ (yes, almost 200 items) while keeping _All->Main.Core root as "interpret"_, I am able to jump inside functions with the debugger. Note that All->Main.Core.Vararg is kept to interpreted, otherwise it will give me the above error.
2. But if I set All->Main.Core root to "compile" then the debugger can no longer jump inside functions.

Is the usage of _Vararg_ in debugger the culprit behind this?

**Comment by MariusDrulea** [comment]
A similar error occurs while setting Main.Core.Vararg to "interpret" this time, while keeping the root Main.Core as "compile".
```
┌ Warning: Setting Vararg as an interpreted method failed.
└ @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:144
┌ Error: Some Julia code in the VS Code extension crashed
└ @ VSCodeDebugger c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\error_handler.jl:15
ERROR: TypeError: in Type, in parameter, expected Type, got Vararg{T, N}
Stacktrace:
  [1] signature_type(f::Any, args::Any)
    @ Base .\reflection.jl:793
  [2] _methods
    @ .\reflection.jl:856 [inlined]
  [3] methods(f::Any, t::Any, mod::Nothing)
    @ Base .\reflection.jl:912
  [4] methods (repeats 2 times)
    @ .\reflection.jl:935 [inlined]
  [5] set_compiled_functions_modules!(items::Vector{String})
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:157
  [6] set_compiled_functions_modules!
    @ c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:173 [inlined]
  [7] set_compiled_items_request(conn::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, state::VSCodeDebugger.DebugAdapter.DebuggerState, params::NamedTuple{(:compiledModulesOrFunctions,), Tuple{Vector{String}}})
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:99
  [8] debug_notification(conn::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, state::VSCodeDebugger.DebugAdapter.DebuggerState, params::VSCodeDebugger.DebugAdapter.DebugArguments)
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\debugger_requests.jl:41
  [9] (::VSCodeDebugger.DebugAdapter.var"#120#146"{VSCodeDebugger.DebugAdapter.DebuggerState})(conn::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, params::VSCodeDebugger.DebugAdapter.DebugArguments)
    @ VSCodeDebugger.DebugAdapter c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\packagedef.jl:49
 [10] dispatch_msg(x::VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, dispatcher::VSCodeDebugger.JSONRPC.MsgDispatcher, msg::Dict{String, Any})
    @ VSCodeDebugger.JSONRPC c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\JSONRPC\src\typed.jl:67    
 [11] macro expansion
    @ c:\Users\some_user_name\.vscode\extensions\julialang.language-julia-1.4.3\scripts\packages\DebugAdapter\src\packagedef.jl:76 [inlined]       
 [12] (::VSCodeDebugger.DebugAdapter.var"#143#169"{VSCodeDebugger.var"#3#4"{Tuple{String, String}}, VSCodeDebugger.JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, VSCodeDebugger.JSONRPC.MsgDispatcher})()
    @ VSCodeDebugger.DebugAdapter .\task.jl:411

Julia debuggee finished. Press ENTER to close this terminal.
```

**Comment by MariusDrulea** [comment]
I have managed to find an acceptable solution to the issue by adding the following lines into the config file. Notice that `Main.Core` is present but `Main.Core.` (dot) is missing. The meaning is: `Main.Core` and all it's symbols (Int8, Arrays etc) are set to compile, but we don't want to apply this rule recursivelly to all sub-modules. I have explicitly set the submodules of "Main.Core" to compile. I have also attached my complete config file.

```
"Main.Core",
"Main.Core.Compiler",
"Main.Core.Compiler.",
"Main.Core.IR",
"Main.Core.IR.",
"Main.Core.Intrinsics",
"Main.Core.Intrinsics."
```
[settings.json.txt](https://github.com/julia-vscode/julia-vscode/files/7226887/settings.json.txt)

**Comment by MariusDrulea** [comment]
@pfitzseb : I have found an acceptable workaround for the debugging problem.
However, the fact that the debugger is able to step inside functions with _Config1_ but not with _Config2_ shall be further investigated in my opinion as this looks potentially buggy.

_Config1_:
```
"Main.Core",
"Main.Core.Compiler",
"Main.Core.Compiler.",
"Main.Core.IR",
"Main.Core.IR.",
"Main.Core.Intrinsics",
"Main.Core.Intrinsics."
```

_Config2_:
```
Main.Core
Main.Core.
```

**Comment by MariusDrulea** [comment]
I have noticed Core contains Main, which further contains Core, which again contains Main etc.
While I understand Main contains Core as I'm using Core built-ins inside Main, I don't see why Core would also contain Main?
And this might explain The issue with ```Main.Core.```, cause this expression would set ```Main.Core.Main``` to compile. 

<img src="https://user-images.githubusercontent.com/27725572/134861130-2f0994ec-cb2b-4f15-9ca1-420b04f85e5e.png" height=500>

**Comment by MariusDrulea** [comment]
And to test my above assumption I have used the following in the config file. I have explicitly excluded ```Main.Core.Main``` from the compile list, and now debugging works fine. 
```
"Main.Core",
"Main.Core.",
"-Main.Core.Main"   
```
